### PR TITLE
qgm: add type inference support 

### LINKED ISSUE #9601: Add support for type inference to QGM
This includes:

- A QGM derived attribute that depends on `NonNullRequirements` with some tests.
- An option to include column types in dot graphs.

Depends on #10409.

### PR DESCRIPTION
**Pull request by aalexandrov**
Resolves #9601.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

* The first commit removes the (now obsolete) `is_div_mod` modifier in the type inference logic for `BinaryFunc`. This is part of [#10964](10964) and will be merged there.
* The second commit adds a new `RelationType` attribute that infers and annotates QGM boxes with a vector of `ColumnType` values corresponding to their output `columns`.
* The last commit adds support for `EXPLAIN [OPTIMIZED] TYPED QUERY GRAPH` statements.
* Fixup commits address some of the issues identified by @philip-stoev and will be squashed in the second commit before I merge the PR.
* ~The fourth commit adds support for exporting typed plans in tests and a `typing` test to validate the new functionality.~ I opted against doing this now, as @philip-stoev will add end-to-end `*.slt` tests akin to the ones in [#10972](https://github.com/MaterializeInc/materialize/pull/10972) tomorrow.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

**Review by sploiselle — APPROVED** [review]
great catch on this; lgtm

**Comment by philip-stoev** [comment]
Item No 1. The QGM path does not detect that this expression is not null:

```
EXPLAIN TYPED QUERY GRAPH FOR SELECT 1 IN ( 1 , NULL   );
```

While the non-QGM path detects that this can never be null. So some constant-folding is happening in the non-QGM case?

The QGM case has the expression expanded like this:

```
boolean? (((false \|\| (1 = 1)) \|\| (1 = null))) }
```
Which I guess can be constant-folded before the type is inferred.

**Comment by philip-stoev** [comment]
Item No2. Same for `CASE`:

```
EXPLAIN TYPED PLAN FOR SELECT  CASE WHEN NULL IS not NULL THEN NULL ELSE 1 END;
```

**Comment by philip-stoev** [comment]
item no 3. Aggregates are inferred incorrectly with QGM:

```
explain typed query graph for SELECT MIN( f2 ) FROM t1;
```

Even though f2 is NOT NULL, MIN over NOT NULL can be NULL.

**Comment by philip-stoev** [comment]
Item No 4. `COALESCE` is inferred incorrectly with QGM:

```
EXPLAIN TYPED QUERY GRAPH FOR SELECT COALESCE( f2 , NULL )  FROM t1;
```

If `f2` is `NOT NULL` then `COALESCE` can never be null, regardless of the nullability of the second argument.

**Comment by philip-stoev** [comment]
Item No5 . `NULLIF` is inferred incorrectly in QGM:

```
SELECT NULLIF( f2 , NULL )
```

If `f2` is NOT NULL, then `f2 = NULL` is guaranteed to be `NULL/FALSE` , so `NULLIF` can not return `NULL` as per its definition.

**Comment by philip-stoev** [comment]
Item No6. `GREATEST/LEAST` is inferred incorrectly in QGM:

```
explain typed query graph  for SELECT greatest( f2 , f2 ) FROM t1
```

As long as **any** of the arguments of those functions is `NOT NULL`, the entire thing can never be NULL.

**Comment by philip-stoev** [comment]
Item No 7. This one is cosmetic so I am putting it here just for completeness. In case the cast is impossible, non-QGM will assign the resulting value a `NOT NULL` property even if it never has the intention of actually returning it.

```
EXPLAIN TYPED PLAN FOR SELECT CAST(CAST('TRUE' AS TEXT) AS BIGINT) FROM t1;
```

QGM on the other hand will not constant-fold the expression so it will not know that should return an error. That said, this is really cosmetic, it is difficult to come up with a scenario where it would matter. It has to have:
- LIMIT 0 so that the error is not thrown at all
- a database client that is looking for NULL/NOT NULL metadata on the (empty) result set.
- 
It does make randomized tests groak periodically though, so if it is a one line fix, it would be nice to have it in.

**Comment by philip-stoev** [comment]
Item No 8.  This query:

```
EXPLAIN TYPED QUERY GRAPH  FOR SELECT f1 NOT IN ( SELECT f1 FROM t1 ) FROM t1
```

results in:
```
2022-03-01T12:23:59.243598Z ERROR panic: attribute mz_sql::query_model::attribute::rejected_nulls::RejectedNulls not present
```

**Comment by philip-stoev** [comment]
Item No 9. Scalar subqueries are not inferred correctly:

```
 EXPLAIN TYPED QUERY GRAPH FOR SELECT ( SELECT 1 FROM t1 ) FROM t2;
```

This can be NULL if t1 is empty, but QGM considers this to be a `integer` result without a `?`

**Comment by aalexandrov** [comment]
> Item No 1. The QGM path does not detect that this expression is not null:
>
> ```sql
> EXPLAIN TYPED QUERY GRAPH FOR SELECT 1 IN ( 1 , NULL );
> ```

This is expanded as `false OR (1 = 1) OR (1 = NULL)`, and without partially evaluating the expression we cannot infer that the `NULL` from the third disjunct will not be propagated to the top. Type inference still works as expected if we remove the `NULL`:

```sql
EXPLAIN TYPED QUERY GRAPH FOR SELECT 1 IN ( 1 , 2 );
```

so I expect that this will be covered in the `EXPLAIN TYPED OPTIMIZED QUERY GRAPH` once we have partial evaluation for  static sub-graphs in place.

**Comment by aalexandrov** [comment]
> Item No2. Same for `CASE`:
> 
> ```sql
> EXPLAIN TYPED PLAN FOR SELECT  CASE WHEN NULL IS not NULL THEN NULL ELSE 1 END;
> ```

Same as above. Observe that for the `TYPED DECORRELATED PLAN` the inference in the current (non-QGM) path is also conservative, as this works on the plan before partial evaluation.

**Comment by aalexandrov** [comment]
> item no 3. Aggregates are inferred incorrectly with QGM:
> 
> ```sql
> explain typed query graph for SELECT MIN( f2 ) FROM t1;
> ```
> 
> Even though f2 is NOT NULL, MIN over NOT NULL can be NULL.

Indeed, there should be an extra case that widens the type of all global aggregates in `Grouping` boxes. I will fix this before we merge this. Great finding, thanks!

**Comment by aalexandrov** [comment]
> Item No 4. `COALESCE` is inferred incorrectly with QGM:
> 
> ```
> EXPLAIN TYPED QUERY GRAPH FOR SELECT COALESCE( f2 , NULL )  FROM t1;
> ```
> 
> If `f2` is `NOT NULL` then `COALESCE` can never be null, regardless of the nullability of the second argument.

Same as Items 1 and 2. This will be covered in the `TYPED OPTIMIZED QUERY GRAPH` once we have partial evaluation. The `TYPED QUERY GRAPH` is identical to the `TYPED DECORRELATED PLAN`, so I don't consider this an error in the type inference.

**Comment by aalexandrov** [comment]
> Item No5 . `NULLIF` is inferred incorrectly in QGM:
> 
> ```sql
> SELECT NULLIF( f2 , NULL ) FROM t1
> ```
> 
> If `f2` is NOT NULL, then `f2 = NULL` is guaranteed to be `NULL/FALSE` , so `NULLIF` can not return `NULL` as per its definition.

Same as Items 1, 2, and 4. More precise type inference here will be a by-product of partial evaluation.

**Comment by aalexandrov** [comment]
> Item No6. `GREATEST/LEAST` is inferred incorrectly in QGM:
> 
> ```sql
> explain typed query graph  for SELECT greatest( f2 , f2 ) FROM t1
> ```
> 
> As long as **any** of the arguments of those functions is `NOT NULL`, the entire thing can never be NULL.

This is a general limitation of the way type inference is implemented for `BinaryFunc` (and any other function for that manner). We are currently propagating the `IS NULL` flag to the output if any of the inputs has it.

We can certainly annotate functions with additional meta-information to guide the type inference logic whenever this can be more precise (and possibly related analysis passes, such as null propagation and null rejection). For example, in this case, one should have a flag such as `cancels_nulls` or something, with the meaning of:

> A function `f` is said to *cancel `NULL` values* if it is guaranteed to not produce a `NOT NULL` output if *any* of its input arguments is known to be `NOT NULL`.

In addition to `greatest`, `coalesce` is another example of a function that can be annotated as "consuming nulls".

I will open a separate issue for that.

**Comment by aalexandrov** [comment]
> Item No 9. Scalar subqueries are not inferred correctly:
> 
> ```
>  EXPLAIN TYPED QUERY GRAPH FOR SELECT ( SELECT 1 FROM t1 ) FROM t2;
> ```
> 
> This can be NULL if t1 is empty, but QGM considers this to be a `integer` result without a `?`

This is a bug that needs to be fixed before we can merge this.

**Comment by philip-stoev** [comment]
>  there should be an extra case that widens the type of all global aggregates in Grouping boxes

Maybe all except `COUNT`

**Comment by aalexandrov** [comment]
> > there should be an extra case that widens the type of all global aggregates in Grouping boxes
> 
> Maybe all except `COUNT`

Forgot to do this, will need to address it before we merge.

File: src/sql/src/query_model/attribute/relation_type.rs
Comment (wangandi) [inline src/sql/src/query_model/attribute/relation_type.rs:184]: I'm trying to understand why is it that the `RelationType` of the input box of the quantifier has always been derived but the `RelationType` of the parent box of the quantifier might not be fully derived.

Is it that we relying on the following two conditions?
1) the correlated subquery is always to the right of the outer query
2) we will always visit left quantifiers before right quantifiers.

If this is the case, we should put down somewhere (possibly rewrite/mod.rs) that rewrites that run before decorrelation cannot change the order of the quantifiers in a box.

Comment (aalexandrov) [reply src/sql/src/query_model/attribute/relation_type.rs:184]: > I'm trying to understand why is it that the RelationType of the input box of the quantifier has always been derived but the RelationType of the parent box of the quantifier might not be fully derived.

This is because derivation proceeds in a bottom-up manner (post-visit), so the `RelationType` of all children will be derived before the `RelationType` of the parent.

So yes, we are relying on the two conditions that you mention. A `Model` instance that has a quantifier `Qi` that depends on `Qj` where `Qi` appears before `Qj` is invalid. We should rather have a `Validator` that asserts this. If you wish, you can open an issue for that.

File: src/sql/src/query_model/attribute/relation_type.rs
Comment (aalexandrov) [reply src/sql/src/query_model/attribute/relation_type.rs:154]: > I think that it is possible for Q3 to become a scalar quantifier after query rewrites such as merging selects and eliminating join identities.

I am not sure how this will work. In general, I am assuming that correlated `ColumnReference` nodes:

1. Always reference a `Foreach` quantifier.
2. The referenced quantifier precedes the quantifier that denotes the sub-graph where the reference occurs.

Comment (aalexandrov) [reply src/sql/src/query_model/attribute/relation_type.rs:154]: Looking into the code, I think the way `correlation_info` is currently written is not respecting the above constraint, but I feel that this is essential, otherwise we will often have to derive the dependency graph of correlated quantifiers ad-hoc. If we impose it, we can at least assume that the quantifiers are topologically sorted with respect to the "Qx has a column that appears as correlated in Qy" relation.

File: src/sql/src/query_model/attribute/relation_type.rs
Comment (wangandi) [reply src/sql/src/query_model/attribute/relation_type.rs:179]: Validator issue created here: https://github.com/MaterializeInc/materialize/issues/11081

File: src/sql/src/query_model/dot.rs
Comment (wangandi) [reply src/sql/src/query_model/dot.rs:84]: I see. Do `transitive_closure` and `dependency_order` ever run separately?

File: src/sql/src/query_model/attribute/relation_type.rs
Comment (wangandi) [reply src/sql/src/query_model/attribute/relation_type.rs:149]: Hmm, you're probably right that collapsing boxes on the left hand side might only be ok if the RHS was not correlated.

In the [implementation of SelectMerge by Andrés](https://github.com/MaterializeInc/materialize/pull/9689/commits/d13d92fd948c94c95f2db64335897f8e974364a1), it is possible to collapse Box2 into Box0 by removing Q3 and replacing it with the quantifiers in Box2. But that would break the graph because Box6 would still refer to Q3.  (This has been noted in the QGM handover notes so we don't forget.)

I just worry that not fully collapsing boxes would block some query rewrites. Perhaps we could have box collapsing happen only after decorrelation. But this requires more thought.

**Review by wangandi — APPROVED** [review]
Code looks fine. Merging the PR should not be blocked on thinking about on `SelectMerge` needs to handle correlated subqueries.

File: src/sql/src/query_model/attribute/relation_type.rs
Comment (aalexandrov) [reply src/sql/src/query_model/attribute/relation_type.rs:149]: Collapsing boxes is a normalizing transform, so most likely we will need to apply it before applying optimizations that rely on this normalized form and after transforms that break it.

Either way, we need to be careful to conceptualize and represent these things upfront in order to avoid the mistakes from the MIR transform crate. If the result is satisfying we should figure out how to migrate the lessons learned to MIR.

File: src/sql/src/query_model/dot.rs
Comment (aalexandrov) [reply src/sql/src/query_model/dot.rs:84]: No. I was thinking that we can introduce a dedicated `DerivableAttributes` type that can only be produced by a smart constructor that takes a list of attributes wraps the transitively closed version of that list in a dependency order. The argument of `derive_dft_attributes` will be changed to be a `DerivableAttributes` instance. This will ensure that we cannot pass an incomplete version of the attributes and we can hide `transitive_closure` and `dependency_order` inside the `DerivableAttributes` constructor.

**Review by philip-stoev — APPROVED** [review]

**Pull request merged**
